### PR TITLE
Print error and stop the script if mandatory Metadata key is not set

### PR DESCRIPTION
**Pull request by pavel-karatsiuba**
If the mandatory Metadata key is not set then print Error and stop the script

Fixes: #1805

### CODE REVIEW

**Comment by kelson42** [comment]
@pavel-karatsiuba It should not be an exception. An exception is a unamanged unexcpected situation in a software. It should be handled properly. That means as early as possible and A proper error message should be given to the user so he understands what is missing.

**Review by kelson42 — CHANGES_REQUESTED** [review]
I see many weaknesses with your implementation:
* The checking is done pretty late, it get all the page details downloaded before checking this (which is useless)
* I know this is not strictly the topic of this ticket... but have you remarked that if the `Description` comes from the Mediawiki, then nothing stops it to be more than 80 charaters?! I kind of expect that this works fine as well.
* If error, the MWoffliner exits with code `0` which is wrong. Remark that if I a give a wrong `--customZimDescription` on the command line then it fails with exit code `2` (which is correct)
* If there is an error then I see something like this (not only I would just prefer to stop pretty quickly but on the top of this the messages are misleading):
```
[info] [2023-03-12T13:56:24.324Z] Getting JSON from [https://bm.wikipedia.org/w/api.php?action=query&format=json&prop=redirects%7Crevisions%7Ccoordinates&rdlimit=max&rdnamespace=0&colimit=max&rawcontinue=true&generator=allpages&gapfilterredir=nonredirects&gaplimit=max&gapnamespace=0&gapcontinue=Interlingua]
[log] [2023-03-12T13:56:24.967Z] Got [407 / 407] articles chunk from namespace [0]
[info] [2023-03-12T13:56:24.967Z] Getting JSON from [https://bm.wikipedia.org/w/api.php?action=query&format=json&prop=redirects%7Crevisions%7Ccoordinates&rdlimit=max&rdnamespace=0&colimit=max&rawcontinue=true&generator=allpages&gapfilterredir=nonredirects&gaplimit=max&gapnamespace=0&gapcontinue=Nsoro]
[log] [2023-03-12T13:56:25.585Z] Got [413 / 820] articles chunk from namespace [0]
[log] [2023-03-12T13:56:25.586Z] A total of [1206] articles has been found in namespace [0]
[log] [2023-03-12T13:56:25.586Z] Got ArticleIDs in 2.09 seconds
[log] [2023-03-12T13:56:25.586Z] Total articles found in Redis: 1207
[warn] [2023-03-12T13:56:25.586Z] Couldn't find strings file for [bm], falling back to [en]
[log] [2023-03-12T13:56:25.587Z] Doing dump
[log] [2023-03-12T13:56:25.587Z] Writing zim to [/home/kelson/code/mwoffliner/out/wikipedia_bm_all_2023-03.zim]
[error] [2023-03-12T13:56:25.587Z] Metadata "{key}" is required
[log] [2023-03-12T13:56:25.587Z] Finished dump
[log] [2023-03-12T13:56:25.587Z] Closing HTTP agents...
[log] [2023-03-12T13:56:25.588Z] All dumping(s) finished with success.
[log] [2023-03-12T13:56:25.588Z] Flushing Redis DBs
[info] [2023-03-12T13:56:25.588Z] Finished running mwoffliner after [5s]
[log] [2023-03-12T13:56:25.588Z] Exiting with code [0]
[log] [2023-03-12T13:56:25.589Z] Deleting temporary directory [/dev/shm/mwoffliner-1678629383338]
```

You need to rethink the architecture of this. First you need to consider that ZIM metadata come AFAIK either from the mediawiki metadata (one short HTTP request at start) or from the command line. Considering this I could propose to:
* Isolate the string metadata constraints in a form of a const table (mandatory, max length, ...)
* Use this table in the sanitizing of command line arguments (when necessary)
* Isolate the metadata sanitizing in a dedicated module (`mwoffliner.lib.ts` is already quite big)
* Reuse the string metadata constraints (created earlier) to check if all the metadata are correct
* Exit in a similar manner if wrong input via command line or just wrong metadata.

**Comment by pavel-karatsiuba** [comment]
@kelson42 If I add sanitizing the Description as we made it for the command line parameter, then many downloads will fail. Maybe I need to trim the Description to 80 characters?

**Comment by kelson42** [comment]
@pavel-karatsiuba obviously you can... even if is is a super minor improvment.... if an improvment at all (strongly suspect Mediawiki to do it already)

**Comment by kelson42** [comment]
@pavel-karatsiuba Any news here? Please rebase/fix conflict before doing anything else.

**Comment by pavel-karatsiuba** [comment]
@kelson42 Ready to review. With the last change, I fixed the test because the new sanitizing rules do not allow Description higher than 80 characters.

**Review by kelson42 — CHANGES_REQUESTED** [review]
> Use this table in the sanitizing of command line arguments (when necessary)

This is not complete, for example max size of `LongDescription` is not checked

File: src/util/metaData.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+const schema = {
+  type: 'object',
+  properties: {
+    Creator: { type: 'string', minLength: 1 },
+    Description: { type: 'string', maxLength: 80, minLength: 1 },
+    Language: { type: 'string', minLength: 1 },
+    Publisher: { type: 'string', minLength: 1 },
+    Title: { type: 'string', maxLength: 30, minLength: 1 },
+  },
+  required: ['Creator', 'Description', 'Language', 'Publisher', 'Title'],
```
Comment (kelson42) [inline src/util/metaData.ts:15]: Should be handled in properties as a `boolean`

File: src/util/metaData.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  additionalProperties: true,
+}
+
+const validate = ajv.compile(schema)
+
+export const validateMetadata = (metaData) => {
+  const valid = validate(metaData)
```
Comment (kelson42) [inline src/util/metaData.ts:22]: I think wording is `sanitize` not `validate`, not great to create inconsistency in the vocabulary.

File: test/unit/util.test.ts [hunk 3]
```diff
@@ -412,4 +413,49 @@ describe('Utils', () => {
       await expect(extractArticleList('http://valid-wrong-url.com/')).rejects.toThrow('Failed to read articleList from URL: http://valid-wrong-url.com/')
     })
   })
+
+  describe('metaData', () => {
```
Comment (kelson42) [inline test/unit/util.test.ts:417]: Your automated tests don't really test the edge cases with undefined, wrong type, missing values.

**Review by kelson42 — CHANGES_REQUESTED** [review]
Still work left... See https://github.com/openzim/zim-tools/pull/339/files, this is the kind of result I  was expected here as well.

File: src/util/metaData.ts (hunk 1 above)
Comment (pavel-karatsiuba) [reply src/util/metaData.ts]: In this case, I used the Ajv JSON schema validator. For this library, the required fields should be passed as an array.

File: src/util/metaData.ts (hunk 2 above)
Comment (pavel-karatsiuba) [reply src/util/metaData.ts:48]: In this case, we only validate the data with some rules but sanitizing implies that data will be modified with some rules. So I propose not renaming the function.

File: src/util/metaData.ts (hunk 1 above)
Comment (kelson42) [reply src/util/metaData.ts]: OK... Not sure this is the right tooling... but let deal with it.

**Review by kelson42 — APPROVED** [review]
LGTM
